packages/assets-controllers/src/token-prices-service/codefi-v2.ts

```diff
@@ -474,64 +437,149 @@ export class CodefiTokenPricesServiceV2
+          asset.tokenAddress.toLowerCase()
+            ? SPOT_PRICES_SUPPORT_INFO[asset.chainId]
+            : `${caipChainId}/erc20:${asset.tokenAddress.toLowerCase()}`) as CaipAssetType,
+        };
+      })
+      .filter((asset) => asset.assetId);
+
+    if (assetsWithIds.length === 0) {
+      return [];
+    }
+
+    const url = new URL(`${BASE_URL_V3}/spot-prices`);
     url.searchParams.append(
-      'tokenAddresses',
-      [getNativeTokenAddress(chainId), ...tokenAddresses].join(','),
+      'assetIds',
+      assetsWithIds.map((asset) => asset.assetId).join(','),
     );
     url.searchParams.append('vsCurrency', currency);
     url.searchParams.append('includeMarketData', 'true');
 
-    const addressCryptoDataMap: MarketDataByTokenAddress =
-      await this.#policy.execute(() =>
-        handleFetch(url, { headers: { 'Cache-Control': 'no-cache' } }),
-      );
-
-    return [getNativeTokenAddress(chainId), ...tokenAddresses].reduce(
-      (
-        obj: Partial<TokenPricesByTokenAddress<Hex, SupportedCurrency>>,
-        tokenAddress,
-      ) => {
-        // The Price API lowercases both currency and token addresses, so we have
-        // to keep track of them and make sure we return the original versions.
-        const lowercasedTokenAddress =
-          tokenAddress.toLowerCase() as Lowercase<Hex>;
+    const addressCryptoDataMap: {
+      [assetId: CaipAssetType]: Omit<
+        MarketDataDetails,
+        'currency' | 'tokenAddress'
+      >;
+    } = await this.#policy.execute(() =>
+      handleFetch(url, { headers: { 'Cache-Control': 'no-cache' } }),
+    );
 
-        const marketData = addressCryptoDataMap[lowercasedTokenAddress];
+    return assetsWithIds
+      .map((assetWithId) => {
+        const marketData = addressCryptoDataMap[assetWithId.assetId];
 
         if (!marketData) {
-          return obj;
+          return undefined;
         }
 
-        const token: TokenPrice<Hex, SupportedCurrency> = {
-          tokenAddress,
-          currency,
+        return {
           ...marketData,
+          ...assetWithId,
+          currency,
         };
+      })
+      .filter((entry): entry is NonNullable<typeof entry> => Boolean(entry));
+  }
 
-        return {
-          ...obj,
-          [tokenAddress]: token,
-        };
+  async #fetchTokenPricesV2(
+    assets: EvmAssetAddressWithChain<SupportedChainId>[],
+    currency: SupportedCurrency,
+  ): Promise<EvmAssetWithMarketData<SupportedChainId, SupportedCurrency>[]> {
+    const v2SupportedAssets = assets.filter(
+      (asset) => !SUPPORTED_CHAIN_IDS_V3.includes(asset.chainId),
+    );
+
+    const assetsByChainId: Record<SupportedChainId, Hex[]> =
+      v2SupportedAssets.reduce(
+        (acc, { chainId, tokenAddress }) => {
+          (acc[chainId] ??= []).push(tokenAddress);
+          return acc;
+        },
+        {} as Record<SupportedChainId, Hex[]>,
+      );
+
+    const promises = Object.entries(assetsByChainId).map(
```

Comment:
Same here. Are there many promises in parallel here?

If yes, thoughts on using the `reduceInBatchesSerially` or other batching utils so we process certain amount at a time?

Also unsure if we should add a timeout mechanism `safelyExecuteWithTimeout`, for if this API hangs for a long time?